pkg/operator/ceph/object/notification/controller.go

```diff
@@ -0,0 +1,257 @@
+/*
```

Comment:
> There is no actual controller for the CephBucketNotification resource that sees add/update/delete actions from a CephBucketNotification and then performs the required actions there.

from the RGW perspective, there are no topics without buckets. so CBNs don't really exist in ceph - this is why the controller does not watch these issues.
in a different PR, I plan on adding cascade delete/update of notification when CBN's are deleted or updated.
for that, I would need a controller that watches over these values, and using the labels in the OBCs get a list of all impacted OBCs, so we can delete/update them.
adding that should not change the user interface, just make the feature more useable.

> if the CBN doesn't exist, log the fact and skip it (if it doesn't yet exist, the CBN controller will detect the CBN when it's created later)

this is the part different in the implementation. currently, if the CBN does not exist I fail the "reconcile" action and requeue it for retries. since we don't modify the state of the OBC (its is still in "Ready" state), this should work.

> use some sort of mutex lock to make sure another routine won't modify the CBN's related configs at the same time

why is this needed? do we do that for other CRs?

regarding `updateCBN()`:
* the OBC part which is implemented its very similar to the suggested design
* the CBN part will be added to address the cascade delete/update

> (is this step necessary?) using Ceph CLI, determine the buckets configured for this notification
for each bucket

there is no ceph CLI to do that. why is this needed?

regarding topics. currently, there are no mechanisms in ceph to prevent deletions of topics "in use", nor a mechanism for updating notifications that had their topics updated (it is even documented in ceph...)
we can add this functionality, but this would be a "nice to have"

Reply:
I should add some more clarity: I don't have a problem with leaving delete/update features for future PRs. We have done that before (notably with RGW multisite), and it has worked well. I do, however, have a strong aversion to merging code that I don't think could maintain.

Reply:
>  I don't have a problem with leaving delete/update features for future PRs.

in this case the CBN controller will be added when we support updates/deletes of notifications based on updates/deletes of CBNs. 

> The CBN needs a controller because even if it isn't a "real" resource in Ceph, we have made it a "real" resource in Rook.

without the update/delete functionality, there is no code that needs to go into the controller

> when implementation of features primarily follows the Ceph design, usability suffers, as does maintainability.

completely agree. but the implementation here does not follow the RGW interfaces. this is why we have CNB (which don't exist in RGW/ceph).